encrypt/aeadCipher.go

```diff
@@ -0,0 +1,376 @@
+	case 16, 24, 32:
+	default:
+		panic("error salt len")
+	}
+	//FIXME
+	n, err := io.ReadFull(rand.Reader, salt)
+	if err != nil || n != len(salt) {
+		panic("error salt gen")
+	}
+}
+
+// key-derivation function for key pre-handle
+func kdf(password string, keyLen int) []byte {
+	var b, prev []byte
+	h := md5.New()
+	for len(b) < keyLen {
+		h.Write(prev)
+		h.Write([]byte(password))
+		b = h.Sum(b)
+		prev = b[len(b)-h.Size():]
+		h.Reset()
+	}
+	return b[:keyLen]
+}
+
+type aeadCipher struct {
+	psk        []byte
+	salt       []byte
+	readNonce  []byte
+	writeNonce []byte
+	readBuffer []byte
+	saltSize   int
+	keySize    int
+	dataLen    int
+
+	enc     cipher.AEAD
+	dec     cipher.AEAD
+	genator func(key, salt []byte, keylen int) (cipher.AEAD, error)
+}
+
+// NewAEADCipher creates a cipher that can be used in Dial() etc.
+// Use cipher.Copy() to create a new cipher with the same method and password
+// to avoid the cost of repeated cipher initialization.
+func NewAEADCipher(method, password string) (c Cipher, err error) {
+	if password == "" {
+		return nil, ErrEmptyPassword
+	}
+
+	var genator *aeadGenerator
+	genator, ok := aeadCipherMethod[method]
+	if !ok {
+		return nil, ErrUnsupportedMethod
+	}
+
+	cipher := aeadCipher{
+		readNonce:  make([]byte, NonceSize, NonceSize),
+		writeNonce: make([]byte, NonceSize, NonceSize),
+		salt:       make([]byte, genator.saltLen, genator.saltLen),
+		psk:        kdf(password, genator.keyLen), // per handle the passwd
+		keySize:    genator.keyLen,
+		saltSize:   genator.saltLen,
+		genator:    genator.newAEAD,
+		readBuffer: make([]byte, readBufferSize),
+	}
+
+	return &cipher, nil
+}
+
+func (c *aeadCipher) KeySize() int          { return c.keySize }
+func (c *aeadCipher) InitBolckSize() int    { return c.saltSize }
+func (c *aeadCipher) EncryptorInited() bool { return c.enc == nil }
+func (c *aeadCipher) DecryptorInited() bool { return c.dec == nil }
+func (c *aeadCipher) InitEncryptor() ([]byte, error) {
+	if c.salt == nil {
+		c.salt = make([]byte, c.saltSize, c.saltSize)
+	}
+	genSalt(c.salt)
+
+	var err error
+	// gen the aead cipher
+	c.enc, err = c.genator(c.psk, c.salt, c.keySize)
+	return c.salt, err
+}
+
+func (c *aeadCipher) InitDecryptor(salt []byte) (err error) {
+	c.salt = make([]byte, c.saltSize, c.saltSize)
+	copy(c.salt, salt)
+
+	// gen the aead cipher
+	c.dec, err = c.genator(c.psk, c.salt, c.keySize)
+	return err
+}
+
+func (c *aeadCipher) Encrypt(src, dest []byte) (int, error) {
+	if c.EncryptorInited() {
+		return -1, ErrCipherUninitialized
+	}
+	msglen := len(src)
+	srclen := make([]byte, 2)
+
+	destlen := dest[:2+c.enc.Overhead()]                                              // data length layout
+	destdata := dest[2+c.enc.Overhead() : 2+c.enc.Overhead()+msglen+c.enc.Overhead()] // data layout
+
+	// handle the length with bigendian encoding
+	srclen[0], srclen[1] = byte(msglen>>8), byte(msglen)
+
+	// encrypt length
+	c.enc.Seal(destlen[:0], c.writeNonce, srclen[:2], nil)
```

Comment:
the data and the length information has been "seal" separately
So, it is still AEAD?
It looks more like  AE to me

Reply:
nope, nonce should increase according to protocol @dszhengyu

Reply:
I mean, the prototype of AEAD.Seal is 
    Seal(dst, nonce, plaintext, additionalData []byte) []byte
the length should be the additionalData parameter and the  Seal  should only be called once, didn't it?

Reply:
Certainly not. According to the protocol, length and meta data should encrypted separately.

http://shadowsocks.org/en/spec/AEAD-Ciphers.html

Reply:
I was questioning the shadowsocks-AEAD-protocol, 
maybe I should ask this in the SIP007 issue

Reply:
@dszhengyu better open up an new issue